tests/kernel/sched/metairq/src/main.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2019 Demant
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel.h>
+
+K_SEM_DEFINE(metairq_sem, 0, 1);
+K_SEM_DEFINE(coop_sem1, 0, 1);
+K_SEM_DEFINE(coop_sem2, 0, 1);
+
+#define LOOP_CNT 10
+
+volatile int coop_cnt1;
+volatile int coop_cnt2;
+
+void metairq_thread(void)
+{
+	k_sem_take(&metairq_sem, K_FOREVER);
+
+	printk("metairq start\n");
+
+	coop_cnt1 = 0;
+	coop_cnt2 = 0;
+
+	printk("give sem2\n");
+	k_sem_give(&coop_sem2);
+
+	k_sleep(100);
+
+	printk("give sem1\n");
+	k_sem_give(&coop_sem1);
+
+	printk("metairq end\n");
+
+	k_sem_give(&metairq_sem);
+}
+
+void coop_thread1(void)
+{
+	k_sem_take(&coop_sem1, K_FOREVER);
+
+	zassert_equal(coop_cnt1, 0,
+		      "Unexpected cnt1 at start: %d", coop_cnt1);
+	zassert_equal(coop_cnt2, LOOP_CNT,
+		      "Unexpected cnt2 at start: %d", coop_cnt2);
+
+	printk("thread1\n");
+	coop_cnt1++;
+
+	zassert_equal(coop_cnt1, 1,
+		      "Unexpected cnt1 at end: %d", coop_cnt1);
+	zassert_equal(coop_cnt2, LOOP_CNT,
+		      "Unexpected cnt2 at end: %d", coop_cnt2);
+
+	k_sem_give(&coop_sem1);
+}
+
+void coop_thread2(void)
+{
+	k_sem_take(&coop_sem2, K_FOREVER);
+
+	zassert_equal(coop_cnt1, 0,
+		      "Unexpected cnt1 at start: %d", coop_cnt1);
+	zassert_equal(coop_cnt2, 0,
+		      "Unexpected cnt2 at start: %d", coop_cnt2);
+
+	for (int i = 0; i < LOOP_CNT; i++) {
+		printk("thread2\n");
+		coop_cnt2++;
+		k_busy_wait(100000);
```

Comment:
Nit: You could reduce this time a bit (to speed up the test and save CI time, now it takes 1 second to run it)

Reply:
Reduced the time an loop count